### PR TITLE
fixed flop counter formula for conv transposed backwards pass

### LINKED ISSUE #119806: FLOP Counter utility seems incorrect for ConvTranspose2d backward
### 🐛 Describe the bug

The flop counter utility at https://github.com/pytorch/pytorch/blob/release/2.1/torch/utils/flop_counter.py
is very useful. Thanks @Chillee for adding it. However, the flops calculation for Conv Transpose backward seems to wildly overestimate the flops.

Example script I ran on H100:
```python
import time
import torch
from torch.utils.flop_counter import FlopCounterMode


if __name__ == "__main__":
    x = torch.rand(6, 192, 30, 96, device="cuda")
    model = torch.nn.ConvTranspose2d(192, 384, (2, 2), stride=2).to("cuda")

    with FlopCounterMode(model):
        start = time.time()
        for i in range(500):
            out = model(x)
            out.sum().backward()

        torch.cuda.synchronize()
        print(f"Elapsed: {time.time() - start}")
```

The output is:
```
Elapsed: 0.5621950626373291
Module                                FLOP    % Total
----------------------------  ------------  ---------
ConvTranspose2d               3677792.061T  25050.00%
 - aten.convolution              1276.568T      8.69%
 - aten.convolution_backward  3676515.493T  25041.31%
 ```
Which implies the loop had 3,677,792.061T / 0.562 =  6,544,113 TFLOPs / s, which is much higher than the peak FLOPs /s on an H100 GPU.

### Versions

Pytorch 2.1

cc @albanD

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119874
* #119521

Fixes #119806

### CODE REVIEW

File: torch/utils/flop_counter.py [hunk 1]
```diff
@@ -95,22 +95,30 @@ def conv_flop_count(
...
+    `prod(conv_shape) * prod(filter_size)` ops). Then, this gets multiplied by
+    1. batch_size, 2. the cross product of input and weight channels.
 
+    For the transpose, it's not each point in the *output* spatial dimension but
+    each point in the *input* spatial dimension.
+    """
     # NB(chilli): I don't think this properly accounts for padding :think:
     # NB(chilli): Should be 2 * c_in - 1 technically for FLOPs.
-    flop = batch_size * prod(conv_shape) * c_out * prod(dims) * 2 * c_in
+    flop = prod(conv_shape) * prod(filter_size) * batch_size * c_out * c_in * 2
```
Comment (zou3519) [inline torch/utils/flop_counter.py:114]: Er, what changed? This looks the same as before

File: torch/utils/flop_counter.py [hunk 2]
```diff
@@ -345,6 +434,9 @@ def _create_pre_module(self, name):
         class PopState(torch.autograd.Function):
             @staticmethod
             def forward(ctx, *args):
+                if self.in_backward:
+                    self.parents = ["Global"]
+                    self.in_backward = True
```
Comment (zou3519) [inline torch/utils/flop_counter.py:439]: What's going on here? I get the changes to the conv flop formula, but not what else is happening

**Review by zou3519 — APPROVED** [review]
seems reasonable

File: torch/utils/flop_counter.py (hunk 1 above)
Comment (Chillee) [reply torch/utils/flop_counter.py:114]: Just some reordering of the formula to match my explanation above.

File: torch/utils/flop_counter.py (hunk 2 above)
Comment (Chillee) [reply torch/utils/flop_counter.py:439]: There was another bug related to clearing of the forwards/backwards stack.

The problem here is that when we have `mod(x)`, we insert an autograd.Function that runs before we enter the module and after we exit the module. Then, in the backwards pass, the corresponding backwards of the autograd.Function runs.

The problem is that if the input does not require grad, it's possible that the backward never runs.

**Pull request closed without merging**